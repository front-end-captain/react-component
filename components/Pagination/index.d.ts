import * as React from "react";

declare type PagerType = "page" | "prev" | "next" | "jumper-prev" | "jumper-next";

export interface PaginationProps {
  // 当前那一页
  current?: number;

  // 默认当前那一页 默认为 1
  defaultCurrent?: number;

  // 每一页条数
  pageSize?: number;

  // 默认每一页条数 默认为 10
  defaultPageSize?: number;

  // 数据总条数
  total?: number;

  // 页面后改变后的回掉 参数是改变后的页码和每页条数
  onChange?: (page: number, pageSize: number) => void;

  // 只有一页时，是否隐藏分页器 默认 FALSE
  hideOnSinglePage?: boolean;

  // TODO 自定义页码的结构
  // itemRender?: (current: number, page: number, type: PagerType, originalElement: React.ReactNode) => React.ReactNode;

  // 用于显示数据总量和当前数据顺序
  showTotal?: (total: number, range: [number, number]) => string;

  // 是否可以快速跳转至某页 默认 FALSE
  showQuickJumper?: boolean;
}

export default class Pagination extends React.Component<PaginationProps, {}> {}
