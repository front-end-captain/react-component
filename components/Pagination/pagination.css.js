import styled from "styled-components";

import { primaryColor } from "./../_style/vars.js";

const PaginationWrapper = styled.ul`
  display: flex;
  flex-direction: row;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  color: rgba(0, 0, 0, 0.65);
  font-size: 14px;
  line-height: 1.5;
  list-style: none;

  &.${(props) => props.prefixCls}-wrapper-single {
    display: none;
  }

  .${(props) => props.prefixCls}-item {
    display: flex;
    justify-content: center;
    align-items: center;
    min-width: 32px;
    height: 32px;
    margin-right: 8px;
    list-style: none;
    background-color: #fff;
    border: 1px solid #d9d9d9;
    border-radius: 4px;
    outline: 0;
    cursor: pointer;
    user-select: none;

    a {
      display: flex;
      width: 100%;
      height: 100%;
      justify-content: center;
      align-items: center;
      color: rgba(0, 0, 0, 0.65);
    }
  }

  .${(props) => props.prefixCls}-item-prev, .${(props) => props.prefixCls}-item-next {
    a {
      font-size: 18px;
    }
  }

  .${(props) => props.prefixCls}-item-selected {
    font-weight: 500;
    border-color: #1890ff;

    a {
      color: #1890ff;
    }
  }

  .${(props) => props.prefixCls}-item-disabled {
    cursor: not-allowed;

    a {
      color: rgba(0, 0, 0, 0.25);
      border-color: #d9d9d9;
      cursor: not-allowed;
    }
  }

  .${(props) => props.prefixCls}-item-jump {
    border: none;
    position: relative;

    a {
      color: rgba(0, 0, 0, 0.35);
      font-size: 18px;

      > span {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        transition: all 0.1s;
      }

      .${(props) => props.prefixCls}-item-jump-arrow {
        opacity: 0;
      }

      .${(props) => props.prefixCls}-item-ellipsis {
        opacity: 1;
      }
    }
  }

  .${(props) => props.prefixCls}-item-jump:hover {
    a {
      color: ${primaryColor};
      font-size: 14px;

      .${(props) => props.prefixCls}-item-jump-arrow {
        opacity: 1;
      }

      .${(props) => props.prefixCls}-item-ellipsis {
        opacity: 0;
      }
    }
  }

  .${(props) => props.prefixCls}-item-total {
    border: none;
  }

  .${(props) => props.prefixCls}-item-quick-jumper {
    border: none;
    margin-left: 15px;

    input {
      position: relative;
      display: inline-block;
      height: auto;
      padding: 4px 11px;
      color: rgba(0, 0, 0, 0.65);
      font-size: 14px;
      line-height: 1.5;
      background-color: #fff;
      background-image: none;
      border: 1px solid #d9d9d9;
      border-radius: 4px;
      transition: all 0.3s;
      width: 30px;
      margin: 0 8px;
    }

    input:focus {
      border-color: #40a9ff;
      border-right-width: 1px !important;
      outline: 0;
      box-shadow: 0 0 0 2px rgba(24, 144, 255, 0.2);
    }
  }
`;

export default PaginationWrapper;
