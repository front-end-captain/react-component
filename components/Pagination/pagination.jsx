import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { ArrowRightIcon, ArrowLeftIcon } from "./../Icon/index.jsx";
import PaginationWrapper from "./pagination.css.js";

class Pagination extends Component {
  static defaultProps = {
    prefixCls: "sky-pagination",
    defaultCurrent: 1,
    defaultPageSize: 10,
    onChange: () => {},
    hideOnSinglePage: false,
    showQuickJumper: false,
  };

  static propTypes = {
    prefixCls: PropTypes.string,
    className: PropTypes.string,
    style: PropTypes.objectOf(PropTypes.string),
    current: PropTypes.number,
    defaultCurrent: PropTypes.number,
    pageSize: PropTypes.number,
    defaultPageSize: PropTypes.number,
    total: PropTypes.number.isRequired,
    onChange: PropTypes.func,
    hideOnSinglePage: PropTypes.bool,
    showTotal: PropTypes.func,
    showQuickJumper: PropTypes.bool,
  };

  constructor(props) {
    super(props);

    this.state = {
      current: props.current || props.defaultCurrent,
      pageSize: props.pageSize || props.defaultPageSize,
    };
  }

  get pageCount() {
    const { total } = this.props;
    const { pageSize } = this.state;
    return Math.ceil(total / pageSize);
  }

  get current() {
    const { current } = this.state;
    return Math.min(current, this.pageCount);
  }

  handleQuickJumper = (event) => {
    const {
      target: { value: jumpPage },
      keyCode,
    } = event;

    if (!jumpPage) {
      return;
    }

    if (keyCode === 13) {
      this.onPageChange(Number(jumpPage));
      this.quickJumperInput.value = "";
    }
  };

  onPageChange = (page) => {
    const { onChange } = this.props;
    const { pageSize } = this.state;
    let realPage = page;

    if (page < 1) {
      realPage = 1;
    }

    if (page >= this.pageCount) {
      realPage = this.pageCount;
    }

    this.setState({ current: realPage }, () => {
      onChange(realPage, pageSize);
    });
  };

  renderPager = (page, current) => {
    const { prefixCls } = this.props;

    const child = <a onClick={() => this.onPageChange(page)}>{page}</a>;

    const element = (
      <li
        key={page}
        title={page}
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-${page}`, {
          [`${prefixCls}-item-selected`]: page === current,
        })}
      >
        {child}
      </li>
    );

    return {
      current,
      page,
      element,
      type: "page",
    };
  };

  renderFirstPager = () => {
    const { prefixCls } = this.props;

    const child = <a onClick={() => this.onPageChange(1)}>1</a>;

    const element = (
      <li key="1" title="1" className={classNames(`${prefixCls}-item`, `${prefixCls}-item-1`)}>
        {child}
      </li>
    );

    return {
      current: this.current,
      page: 1,
      element,
      type: "page",
    };
  };

  renderLastPager = (pageCount) => {
    const { prefixCls } = this.props;

    const child = <a onClick={() => this.onPageChange(pageCount)}>{pageCount}</a>;

    const element = (
      <li
        key={pageCount}
        title={pageCount}
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-${pageCount}`)}
      >
        {child}
      </li>
    );

    return {
      current: this.current,
      page: pageCount,
      element,
      type: "pager",
    };
  };

  renderPrevPageBtn = (current) => {
    const { prefixCls } = this.props;
    const child = (
      <a onClick={() => this.onPageChange(current - 1)}>
        <ArrowLeftIcon />
      </a>
    );

    const element = (
      <li
        title="上一页"
        key="prev"
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-prev`, {
          [`${prefixCls}-item-disabled`]: current === 1,
        })}
      >
        {child}
      </li>
    );

    return {
      current,
      page: null,
      element,
      type: "prev",
    };
  };

  renderNextPageBtn = (current, pageCount) => {
    const { prefixCls } = this.props;
    const child = (
      <a onClick={() => this.onPageChange(current + 1)}>
        <ArrowRightIcon />
      </a>
    );

    const element = (
      <li
        title="下一页"
        key="next"
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-next`, {
          [`${prefixCls}-item-disabled`]: current === pageCount,
        })}
      >
        {child}
      </li>
    );

    return {
      current,
      page: null,
      element,
      type: "next",
    };
  };

  renderJumpPrev = (current) => {
    const { prefixCls } = this.props;
    const child = (
      <a onClick={() => this.onPageChange(current - 5)}>
        <span className={`${prefixCls}-item-ellipsis`}>···</span>
        <span className={`${prefixCls}-item-jump-arrow`}>{"<<"}</span>
      </a>
    );

    const element = (
      <li
        key="jump-prev"
        title="向前5页"
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-jump`)}
      >
        {child}
      </li>
    );

    return {
      current,
      page: null,
      element,
      type: "jumper-prev",
    };
  };

  renderJumpNext = (current) => {
    const { prefixCls } = this.props;
    const child = (
      <a onClick={() => this.onPageChange(current + 5)}>
        <span className={`${prefixCls}-item-ellipsis`}>···</span>
        <span className={`${prefixCls}-item-jump-arrow`}>{">>"}</span>
      </a>
    );

    const element = (
      <li
        key="jump-next"
        title="向后5页"
        className={classNames(`${prefixCls}-item`, `${prefixCls}-item-jump`)}
      >
        {child}
      </li>
    );

    return {
      current,
      page: null,
      element,
      type: "jumper-next",
    };
  };

  renderPagerList = () => {
    const pagerList = [];
    const {
      pageCount,
      state: { current: currentPage },
    } = this;

    if (pageCount <= 9) {
      for (let i = 1; i <= pageCount; i += 1) {
        pagerList.push(this.renderPager(i, currentPage).element);
      }
    } else {
      let left = Math.max(1, currentPage - 2);
      let right = Math.min(currentPage + 2, pageCount);

      if (currentPage - 1 <= 2) {
        right = 1 + 4;
      }

      if (pageCount - currentPage <= 2) {
        left = pageCount - 4;
      }

      for (let j = left; j <= right; j += 1) {
        pagerList.push(this.renderPager(j, currentPage).element);
      }

      if (currentPage - 1 >= 4) {
        pagerList.unshift(this.renderJumpPrev(currentPage).element);
      }

      if (pageCount - currentPage >= 4) {
        pagerList.push(this.renderJumpNext(currentPage).element);
      }

      if (left !== 1) {
        pagerList.unshift(this.renderFirstPager().element);
      }

      if (right !== currentPage && right !== pageCount) {
        pagerList.push(this.renderLastPager(pageCount).element);
      }
    }

    pagerList.unshift(this.renderPrevPageBtn(currentPage).element);
    pagerList.push(this.renderNextPageBtn(currentPage, pageCount).element);

    return pagerList;
  };

  renderTotal = () => {
    const { showTotal, prefixCls, total } = this.props;
    const { pageSize } = this.state;
    const range = [pageSize * this.current - (pageSize - 1), pageSize * this.current];

    if (showTotal && typeof showTotal === "function") {
      return (
        <li className={classNames(`${prefixCls}-item`, `${prefixCls}-item-total`)}>
          {showTotal(total, range)}
        </li>
      );
    }

    return null;
  };

  renderQuickJumper = () => {
    const { prefixCls, showQuickJumper } = this.props;

    if (showQuickJumper) {
      return (
        <li className={classNames(`${prefixCls}-item`, `${prefixCls}-item-quick-jumper`)}>
          <span>跳至</span>
          <input
            type="text"
            onKeyDown={this.handleQuickJumper}
            ref={(node) => (this.quickJumperInput = node)}
          />
          <span>页</span>
        </li>
      );
    }

    return null;
  };

  render() {
    const { prefixCls, className, style, hideOnSinglePage } = this.props;
    const paginationWrapperClassName = classNames(`${prefixCls}-wrapper`, className, {
      [`${prefixCls}-wrapper-single`]: hideOnSinglePage && this.pageCount === 1,
    });

    return (
      <PaginationWrapper prefixCls={prefixCls} style={style} className={paginationWrapperClassName}>
        {this.renderTotal()}
        {this.renderPagerList()}
        {this.renderQuickJumper()}
      </PaginationWrapper>
    );
  }
}

export default Pagination;
