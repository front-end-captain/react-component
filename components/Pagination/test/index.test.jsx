import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Pagination from "./../index.js";

describe("Pagination", () => {
  it("should render some Pagination components", () => {
    const wrapper = render(
      <div>
        <Pagination total={90} />
        <Pagination total={5000} defaultPageSize={5} defaultCurrent={7} />
        <Pagination
          total={400}
          showTotal={(total, range) => `${range[0]}-${range[1]} of ${total} items`}
        />
        <Pagination total={1000} defaultPageSize={5} showQuickJumper />
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
