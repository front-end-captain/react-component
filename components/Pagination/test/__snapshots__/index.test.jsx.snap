// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should render some Pagination components 1`] = `
<div>
  <ul
    class="sky-pagination-wrapper paginationcss__PaginationWrapper-vesd20-0 fOSaqI"
  >
    <li
      class="sky-pagination-item sky-pagination-item-prev sky-pagination-item-disabled"
      title="上一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-1 sky-pagination-item-selected"
      title="1"
    >
      <a>
        1
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-2"
      title="2"
    >
      <a>
        2
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-3"
      title="3"
    >
      <a>
        3
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-4"
      title="4"
    >
      <a>
        4
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-5"
      title="5"
    >
      <a>
        5
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-6"
      title="6"
    >
      <a>
        6
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-7"
      title="7"
    >
      <a>
        7
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-8"
      title="8"
    >
      <a>
        8
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-9"
      title="9"
    >
      <a>
        9
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-next"
      title="下一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </a>
    </li>
  </ul>
  <ul
    class="sky-pagination-wrapper paginationcss__PaginationWrapper-vesd20-0 fOSaqI"
  >
    <li
      class="sky-pagination-item sky-pagination-item-prev"
      title="上一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-1"
      title="1"
    >
      <a>
        1
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-jump"
      title="向前5页"
    >
      <a>
        <span
          class="sky-pagination-item-ellipsis"
        >
          ···
        </span>
        <span
          class="sky-pagination-item-jump-arrow"
        >
          &lt;&lt;
        </span>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-5"
      title="5"
    >
      <a>
        5
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-6"
      title="6"
    >
      <a>
        6
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-7 sky-pagination-item-selected"
      title="7"
    >
      <a>
        7
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-8"
      title="8"
    >
      <a>
        8
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-9"
      title="9"
    >
      <a>
        9
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-jump"
      title="向后5页"
    >
      <a>
        <span
          class="sky-pagination-item-ellipsis"
        >
          ···
        </span>
        <span
          class="sky-pagination-item-jump-arrow"
        >
          &gt;&gt;
        </span>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-1000"
      title="1000"
    >
      <a>
        1000
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-next"
      title="下一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </a>
    </li>
  </ul>
  <ul
    class="sky-pagination-wrapper paginationcss__PaginationWrapper-vesd20-0 fOSaqI"
  >
    <li
      class="sky-pagination-item sky-pagination-item-total"
    >
      1-10 of 400 items
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-prev sky-pagination-item-disabled"
      title="上一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-1 sky-pagination-item-selected"
      title="1"
    >
      <a>
        1
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-2"
      title="2"
    >
      <a>
        2
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-3"
      title="3"
    >
      <a>
        3
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-4"
      title="4"
    >
      <a>
        4
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-5"
      title="5"
    >
      <a>
        5
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-jump"
      title="向后5页"
    >
      <a>
        <span
          class="sky-pagination-item-ellipsis"
        >
          ···
        </span>
        <span
          class="sky-pagination-item-jump-arrow"
        >
          &gt;&gt;
        </span>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-40"
      title="40"
    >
      <a>
        40
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-next"
      title="下一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </a>
    </li>
  </ul>
  <ul
    class="sky-pagination-wrapper paginationcss__PaginationWrapper-vesd20-0 fOSaqI"
  >
    <li
      class="sky-pagination-item sky-pagination-item-prev sky-pagination-item-disabled"
      title="上一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="15 18 9 12 15 6"
          />
        </svg>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-1 sky-pagination-item-selected"
      title="1"
    >
      <a>
        1
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-2"
      title="2"
    >
      <a>
        2
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-3"
      title="3"
    >
      <a>
        3
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-4"
      title="4"
    >
      <a>
        4
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-5"
      title="5"
    >
      <a>
        5
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-jump"
      title="向后5页"
    >
      <a>
        <span
          class="sky-pagination-item-ellipsis"
        >
          ···
        </span>
        <span
          class="sky-pagination-item-jump-arrow"
        >
          &gt;&gt;
        </span>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-200"
      title="200"
    >
      <a>
        200
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-next"
      title="下一页"
    >
      <a>
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1em"
        >
          <polyline
            points="9 18 15 12 9 6"
          />
        </svg>
      </a>
    </li>
    <li
      class="sky-pagination-item sky-pagination-item-quick-jumper"
    >
      <span>
        跳至
      </span>
      <input
        type="text"
      />
      <span>
        页
      </span>
    </li>
  </ul>
</div>
`;
